ietf/submit/views.py

```diff
@@ -106,6 +109,105 @@ def upload_submission(request):
                               {'selected': 'index',
                                'form': form})
 
+@csrf_exempt
+def api_upload(request):
+    def err(code, error, messages=None):
+        data = {'error': error}
+        if messages is not None:
+            data['messages'] = [messages] if isinstance(messages, str) else messages
+        return JsonResponse(data, status=code)
+
+    if request.method == 'GET':
+        return render(request, 'submit/api_submit_info.html')
+    elif request.method == 'POST':
+        exception = None
+        submission = None
+        try:
+            form = SubmissionAutoUploadForm(request, data=request.POST, files=request.FILES)
+            if form.is_valid():
+                log('got valid submission form for %s' % form.filename)
+                username = form.cleaned_data['user']
+                user = User.objects.filter(username=username)
+                if user.count() == 0:
+                    # See if a secondary login was being used
+                    email = Email.objects.filter(address=username, active=True)
+                    # The error messages don't talk about 'email', as the field we're
+                    # looking at is still the 'username' field.
+                    if email.count() == 0:
+                        return err(400, "No such user: %s" % username)
+                    elif email.count() > 1:
+                        return err(500, "Multiple matching accounts for %s" % username)
+                    email = email.first()
+                    if not hasattr(email, 'person'):
+                        return err(400, "No person matches %s" % username)
+                    person = email.person
+                    if not hasattr(person, 'user'):
+                        return err(400, "No user matches: %s" % username)
+                    user = person.user
+                elif user.count() > 1:
+                    return err(500, "Multiple matching accounts for %s" % username)
+                else:
+                    user = user.first()
+                if not hasattr(user, 'person'):
+                    return err(400, "No person with username %s" % username)
+
+                clear_existing_files(form)
+                save_files(form)
+
+                submission = get_submission(form)
+                submission.state = DraftSubmissionStateName.objects.get(slug="validating")
+                submission.remote_ip = form.remote_ip
+                submission.file_types = ','.join(form.file_types)
+                submission.submission_date = datetime.date.today()
+                submission.submitter = user.person.formatted_email()
+                submission.save()
+                create_submission_event(request, submission, desc="Uploaded submission through API")
+
+                # Wrap in on_commit so the delayed task cannot start until the view is done with the DB
+                transaction.on_commit(
```

Comment:
does this need to be moved relative to the save() call?

Reply:
No - it will be called after a hypothetical DB write in the same spot would have been committed, or immediately if there is no transaction in progress. I.e., it needs to be called when we're done changing the DB.